Set up proper registration flow

anoadragon453 opened the pull request:
* We now keep track of sessions and their completed registration stages.
* We only complete registration if the client has completed a full flow.
* New Derived section in config for data derived from config options.
* New config options for captcha.
* Send params back to client for each registration stage.

Required for #343

anoadragon453 commented on src/github.com/matrix-org/dendrite/clientapi/routing/register.go:
```diff
@@ -212,16 +238,36 @@ func Register(
+
+	// There are still more stages to complete.
+	// Return the flows and those that have been completed.
+	return util.JSONResponse{
```
@erikjohnston Did you mention that you typically use `return httputil.LogThenError(req, err)` for this rather than creating a `util.JSONResponse` on the spot?

erikjohnston replied:
Only for when returning 500s. There may be other helper functions for doing 401

erikjohnston replied:
But I think this is fine as is

anoadragon453 replied:
Cool, alright.

erikjohnston commented on src/github.com/matrix-org/dendrite/clientapi/routing/register.go:
```diff
@@ -376,6 +422,62 @@ func isValidMacLogin(
+// RandString returns a random string of characters with a given length.
+// Do note that it is not thread-safe in its current form.
+// https://stackoverflow.com/a/31832326
+func RandString(n int) string {
```
`util.RandomString` is already a thing

anoadragon453 replied:
Cool, was hoping we'd have something already :)

erikjohnston commented on src/github.com/matrix-org/dendrite/clientapi/routing/register.go:
```diff
@@ -376,6 +422,62 @@ func isValidMacLogin(
+	sort.Strings(aCopy)
+	sort.Strings(bCopy)
+
+	return reflect.DeepEqual(aCopy, bCopy)
```
Since these are just slices of strings I'd just do a for loop. I don't trust the reflect package not to be weird or slow.

```golang
for i := range aCopy {
    if a[i] != b[i] {
        return false
    }
}
return true
```

erikjohnston replied:
It might actually be nicer to just do, instead of creating copies:

```golang
sort.Slice(aFlow, func(x, y) { return x < y })
```

anoadragon453 replied:
> It might actually be nicer to just do, instead of creating copies:
>     `sort.Slice(aFlow, func(x, y) { return x < y })`

Oh, much cleaner. I'll give that a whirl.

anoadragon453 replied:
Nice, works: https://play.golang.com/p/wLPjs2gfWE

erikjohnston commented on src/github.com/matrix-org/dendrite/clientapi/routing/register.go:
```diff
@@ -43,9 +46,13 @@ const (
 
-var validUsernameRegex = regexp.MustCompile(`^[0-9a-zA-Z_\-./]+$`)
+var (
+	sessions           = make(map[string][]authtypes.LoginType) // Sessions and completed flow stages
```
We'll want to prune these sessions so that we don't just slowly but surely fill up with old sessions.

Happy for you to just add a TODO and can deal with it in another PR.

anoadragon453 replied:
Yeah, good catch. Obviously removing when registering is completed is fine and dandy, but will need to do something about a session length for non-completed registrations.

erikjohnston replied:
Actually, we'll probably want to keep around completed registrations for a while in case the client retries the last register request.

anoadragon453 replied:
You mean even after `completeRegistration()` has been carried out?

In that case we can just have a session length be the only way to prune sessions.

erikjohnston commented:
(Assigning to @richvdh to double check that this is how registration actually works)

erikjohnston commented on src/github.com/matrix-org/dendrite/common/config/config.go:
```diff
@@ -189,6 +190,17 @@ type Dendrite struct {
+		// Params for registration. Data that is returned to the client while registering and
+		// that which is necessary to complete certain registration flow stages
+		Params map[string]interface{} `json:"params"`
+	}
```
We may want to stick this in a `Registration` sub structure, otherwise its not clear that `config.Derived.Flows` is a registration thing

richvdh replied:
+1.

TBH I'm not entirely sure this belongs in the Config object at all. Shouldn't it be up to the register.go module to figure out the right flows based on the config?

anoadragon453 replied:
The intention was to generate these structures once on startup, then use them over and over again during each registration.

We could place the actual calculation in register.go, though we would simply be calling it from config.go after the config file was read.

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/auth/authtypes/flow.go:
```diff
@@ -0,0 +1,21 @@
+package authtypes
+
+// Flow represents one possible way that the client can authenticate a request.
+// http://matrix.org/speculator/spec/HEAD/client_server/unstable.html#user-interactive-authentication-api
```
suggest linking to https://matrix.org/docs/spec/client_server/r0.3.0.html. Even if you need the unstable version (which you don't here), https://matrix.org/docs/spec/client_server/unstable.html is more reliable than the speculator.

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register.go:
```diff
@@ -376,6 +420,25 @@ func isValidMacLogin(
+func checkFlowsEqual(aFlow, bFlow authtypes.Flow) bool {
+	a := aFlow.Stages
+	b := bFlow.Stages
+	if len(a) != len(b) {
```
it's valid for the client to complete *extra* stages.

richvdh replied:
This is still going to fail if the client completes extra stages.

You're also going to have to rename the function and change its comment.

anoadragon453 commented:
Updated to fix spec links, allow client to complete extra flow stages and still register and create a Registration sub-struct under Derived.

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/auth/authtypes/flow.go:
```diff
@@ -0,0 +1,21 @@
+package authtypes
+
+// Flow represents one possible way that the client can authenticate a request.
+// http://matrix.org/docs/spec/HEAD/client_server/r0.3.0.html#user-interactive-authentication-api
```
broken link

richvdh commented:
Some other points:

* please could you sign off with your full name? Sorry but those are the rules :/
* it would make reviewing changes easier if you made changes in new commits rather than amending the previous one
* it feels like we should have some tests on this stuff

anoadragon453 commented:
I've cleaned up the flow-checking code, switched to my real name (thanks for poking me to do that finally), and added some simple tests for the flow-checking.

We should probably have some integration tests for registration in general though. Do you think those should be bundled in this PR or put in another one?

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register.go:
```diff
@@ -251,21 +251,19 @@ func handleRegistrationFlow(
-			return completeRegistration(req.Context(), accountDB, deviceDB,
-				r.Username, r.Password, r.InitialDisplayName)
+	// Check if the user's registration flow has been completed successfully
+	if !checkFlowCompleted(authtypes.Flow{sessions[sessionID]}, cfg.Derived.Registration.Flows) {
```
it feels odd that we are constructing a `Flow` here and then immediately unpacking it in `checkFlowCompleted`. Why not just pass in the `[]LoginType` ?

anoadragon453 replied:
My inner OCD wanted to only accept Flow's into checkFlowCompleted, but yes I can change it :)

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register.go:
```diff
@@ -420,36 +418,47 @@ func isValidMacLogin(
-	extraStages := len(b) - len(a)
+	// Account for any extra stages a user may unnecessarily do.
+	extraStages := len(a) - len(b)
 	for i := range b {
```
this doesn't look right.

Given:
```
a = { 'A', 'B' }
b = { 'Z' }
```

does this do the right thing (ie, return false)? Please add a test for it.

Also I think you're requiring that flows are completed in the same order as they are configured, which isn't necessary.

anoadragon453 replied:
Since they're sorted, the order in fact does not matter. You can play with the algo here:

https://play.golang.org/p/CHGHI03OBO

I've added the test case to both the playground and the .go file, both work :)

richvdh replied:
```
	derivedFlows = []Flow{Flow{[]LoginType{LoginType("Z")}}}
	fmt.Println("Derived Flows:", derivedFlows)

	userFlow2 := Flow{[]LoginType{LoginType("A"), LoginType("B")}}
	fmt.Println("userFlow2:", userFlow2, checkFlowCompleted(userFlow2))
```

that gives true for me. I don't call that "working".

anoadragon453 replied:
Ahh, misread your original comment and switched the allowed and input flows. 

Yes, that input does result in the incorrect output!

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register_test.go:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	if !checkFlowCompleted(testFlow1, derivedFlows) {
+		t.Error(fmt.Sprint("Failed to verify registration flow: ", testFlow1, ", from derived flows: ", derivedFlows, ". Should be true."))
```
`fmt.Sprint` is unnecessary here

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register_test.go:
```diff
@@ -0,0 +1,86 @@
+)
+
+func TestFlowChecking(t *testing.T) {
+	derivedFlows := []authtypes.Flow{
```
`derivedFlows` seems an odd name here. Should it be `allowedFlows` or something?

anoadragon453 replied:
That is nicer sounding, yes.

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register_test.go:
```diff
@@ -0,0 +1,86 @@
+		t.Error(fmt.Sprint("Failed to verify registration flow: ", testFlow1, ", from derived flows: ", derivedFlows, ". Should be true."))
+	}
+	if checkFlowCompleted(testFlow2, derivedFlows) {
+		t.Error(fmt.Sprint("Failed to verify registration flow: ", testFlow2, ", from derived flows: ", derivedFlows, ". Should be false."))
```
given you're testing that `testFlow2` can *not* be verified against `derivedFlows`, I think this error message is misleading.

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register_test.go:
```diff
@@ -0,0 +1,86 @@
+		},
+	}
+
+	if !checkFlowCompleted(testFlow1, derivedFlows) {
```
I think it would be nice to split each of these tests out into its own function, which is named according to what you're trying to test. You can make `derivedFlows` a global const to make that easier.

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register_test.go:
```diff
@@ -0,0 +1,86 @@
+// Copyright 2017 Vector Creations Ltd
+// Copyright 2017 New Vector Ltd
```
We don't ask you to assign copyright, and neither Vector Creations nor New Vector wrote this code, so the copyright belongs to you. You can either put your own name here or leave out the Copyright lines.

richvdh replied:
(likewise you're free to add an assertion of your own copyright to any files you make non-trivial changes to. Not that I think it matters much from the legal point of view)

anoadragon453 commented:
Should be good to go™

richvdh commented on src/github.com/matrix-org/dendrite/clientapi/routing/register.go:
```diff
@@ -425,18 +425,21 @@ func checkFlows(a []authtypes.LoginType, b []authtypes.LoginType) bool {
 	sort.Slice(a, func(i, j int) bool { return a[i] < a[j] })
```
hrm. This is going to modify the inputs to `checkFlows` - ie, it's going to write to `cfg.Derived.Registration.Flows` and `sessions[sessionID]`.

I guess it's harmless enough, but it's certainly unexpected. It also highlights the redundancy here in sorting for every pairing of session/flow.

My suggestion would be that you do one of:
* sort the lists much earlier and require sorted lists be passed into `checkFlowCompleted`
* rather than sorting the lists, construct a `map[LoginType]bool` for the completed stages and iterate over the steps in `b`

While I'm here, can we come up with better names than `a` and `b`? `completedStages` / `requiredStages` ?

anoadragon453 replied:
Would simply creating a copy of the stages before sorting them in-place be sufficient?

anoadragon453 commented on src/github.com/matrix-org/dendrite/clientapi/routing/register_test.go:
```diff
@@ -109,3 +109,26 @@ func TestFlowCheckingUnorderedAndExtraneous(t *testing.T) {
+	}
+}
+
+// Should return false as we're providing less stages than are required.
```
Whoops, copy/paste error.

anoadragon453 commented:
We no longer unexpectedly sort required and allow slices. Also switched out the `a` and `b` var names for more descriptive ones.

anoadragon453 commented:
Heh, thanks for your own!

On November 29, 2017 1:41:46 AM PST, Richard van der Hoff <notifications@github.com> wrote:
>richvdh approved this pull request.
>
>it feels inefficient to be repeatedly copying and sorting those lists,
>but fine, we can always address that later if it turns out to be an
>issue.
>
>LGTM. Thanks for your perseverence!
>
>
>
>-- 
>You are receiving this because you were assigned.
>Reply to this email directly or view it on GitHub:
>https://github.com/matrix-org/dendrite/pull/354#pullrequestreview-79784438
